File: lib/TwemojiMapper.rb
#######################################################
# TWEMOJI SCRAPER: the source is located at:
# https://twemoji.maxcdn.com/2/test/preview.html
#######################################################

require 'capybara'
require 'capybara/poltergeist'
require 'csv'

require 'pry'

module TwemojiMapper

  include Capybara::DSL
  Capybara.default_driver = :poltergeist

  @URL = "https://twemoji.maxcdn.com/2/test/preview.html"
  @OUTPUT_PATH = File.join(File.expand_path("..", Dir.pwd), "tmp", "twemoji_list.csv")


#######################################################
# FUNCTIONS: for process legibility
#######################################################

  def self.codeOf node

  end


#######################################################
# ANONYMOUS FXNS: for call-chaining legibility
#######################################################

  @pngName = lambda do |node|
    binding.pry
    self.codeOf node
  end



#######################################################
# CALLING THE SCRAPE/MAP PROCESS, WRITING THE LIST
#######################################################

  def self.generateCSV
    puts "generating twemoji code point list"

    puts "loading full twemoji list"
    #doc = Nokogiri::HTML(open(@URL))
    visit @URL

    output = []

    all("li img").each do |twemoji|
      binding.pry
      output << twemoji
    end

    puts "twemoji list loaded, converting to list"
    #twemoji_entries = doc
      #.css("li img")
      #.map(&@pngName)

    puts "list generated, writing csv"
    CSV.open(@OUTPUT_PATH, "wb") do |csv|
      twemoji_entries.each do |entry|
        csv << entry
      end
    end

    puts "csv writing complete"
  end


end

File: lib/TwemojiMapper.js
var fs = require('fs');
var path = require('path');
var jsdom = require('jsdom');

var OUTPUT_PATH = path.join(__dirname, '..', 'tmp', 'twemoji_list.csv');

console.log('starting scrape');

jsdom.env({
  url: 'https://twemoji.maxcdn.com/2/test/preview.html',
  scripts: ['http://code.jquery.com/jquery.js'],
  FetchExternalResources: ['script'],
  ProcessExternalResources: ['script'],
  MutationEvents: '2.0',
  done: function(error, window) {

    $ = window.$;

    console.log($('li img').toArray());

    setTimeout(function() {
      var output = Array
        .prototype
        .slice
        .call(window.document.querySelectorAll('li img'))
        .map(function(el) {
          return el
            .src
            .split('/')
            .slice(-1)[0]
            .split('.')[0];
        })
        .join(/\n/);
      fs.writeFileSync(OUTPUT_PATH, output);
    }, 15000);
  }
});

